test/csuite/wt3363_checkpoint_op_races/main.c

```diff
@@ -0,0 +1,254 @@
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "test_util.h"
+
+/*
+ * JIRA ticket reference: WT-3363
+ *
+ * Test case description: There are a number of operations that we run that we
+ * expect not to conflict with or block against a running checkpoint. This test
+ * aims to run repeated checkpoints in a thread, while running an assortment
+ * of operations that we expect to execute quickly on further threads. To
+ * ensure that we catch any blockages we introduce a very large delay into the
+ * checkpoint and measure that no operation takes 1/2 the length of this delay.
+ *
+ * Failure mode: We monitor the execution time of all operations and if we find
+ * any operation taking longer than 1/2 the delay time, we abort dumping a core
+ * file which can be used to determine what operation was blocked.
+ */
+
+void* do_checkpoints(void *);
+void* do_ops(void *);
+void* monitor(void *);
+
+/*
+ * Time delay to introduce into checkpoints in seconds. Should be at-least
+ * double the maximum time that any one of the operations should take. Currently
+ * this is set to 10 seconds and we expect no single operation to take longer
+ * than 5 seconds.
+ */
+#define	MAX_EXECUTION_TIME 10
+#define	N_THREADS 10
+/*
+ * Number of seconds to execute for. Initially set to 15 minutes, as we need to
+ * run long enough to be certain we have captured any blockages. In initial
+ * testing 5 minutes was enough to reproduce the issue, so we run for 3x that
+ * here to ensure we reproduce before declaring success.
+ */
+#define	RUNTIME 900.0
+
+static WT_EVENT_HANDLER event_handler = {
+	handle_op_error,
+	handle_op_message,
+	NULL,
+	NULL
+};
+
+int
+main(int argc, char *argv[])
+{
+	PER_THREAD_ARGS thread_args[N_THREADS];
+	TEST_OPTS *opts, _opts;
+	pthread_t ckpt_thread, mon_thread, threads[N_THREADS];
+	uint64_t uid;
+	int i;
+
+	if (!testutil_enable_long_tests())	/* Ignore unless requested. */
+		return (EXIT_SUCCESS);
+
+	uid = 0;
+	opts = &_opts;
+	memset(opts, 0, sizeof(*opts));
+
+	testutil_check(testutil_parse_opts(argc, argv, opts));
+	testutil_make_work_dir(opts->home);
+
+	testutil_check(wiredtiger_open(
+	    opts->home, &event_handler, "create,cache_size=1G,", &opts->conn));
+
+	testutil_check(pthread_create(
+	    &ckpt_thread, NULL, do_checkpoints, (void *)opts));
+
+	for (i = 0; i < N_THREADS; ++i) {
+		thread_args[i].uid = &uid;
+		thread_args[i].testopts = opts;
+		thread_args[i].thread_counter = 0;
+		thread_args[i].threadnum = i;
+		testutil_check(pthread_create(
+		    &threads[i], NULL, do_ops, (void *)&thread_args[i]));
+	}
+
+	testutil_check(
+	    pthread_create(&mon_thread, NULL, monitor, &thread_args));
```

Comment:
The parameter being passed to the monitor thread deserves a comment. It's odd that it's different to the other threads.